File: runtime/schema/python/superschematic_schema_runtime/parse/registry.py
"""Scalar parse / normalize registries.

Pulls the default function set from the generated ``_generated_default_registry``
module so the registry composition is identical across calls and codegen runs.
"""

from __future__ import annotations

from typing import Callable

from .. import ValidationError
from .._generated_default_registry import (
    DEFAULT_NORMALIZE_FUNCTIONS,
    DEFAULT_PARSE_FUNCTIONS,
)
from ..validation.types import Schema


ScalarParseFunc = Callable[[str], tuple[str, list[ValidationError]]]
ScalarNormalizeFunc = Callable[[str], str]


def _wrap_parse(fn: Callable[[str], str], label: str) -> ScalarParseFunc:
    """Adapt a ``parse_<name>(str) -> str`` (raises ``ValueError``) to the
    runtime contract that returns ``(value, errors)``."""

    def adapter(input_value: str) -> tuple[str, list[ValidationError]]:
        try:
            return fn(input_value), []
        except ValueError as exc:
            return input_value, [
                ValidationError(validator="parse", message=f"invalid {label}: {exc}")
            ]

    return adapter


class ScalarParseRegistry:
    """Registry of scalar parse functions keyed by canonical name."""

    def __init__(self) -> None:
        self._fns: dict[str, ScalarParseFunc] = {}

    def register(self, name: str, fn: ScalarParseFunc) -> None:
        self._fns[name] = fn

    def unregister(self, name: str) -> bool:
        return self._fns.pop(name, None) is not None

    def get(self, name: str) -> ScalarParseFunc | None:
        return self._fns.get(name)

    def has(self, name: str) -> bool:
        return name in self._fns

    def names(self) -> list[str]:
        return sorted(self._fns)

    def missing_parsers(self, schema: Schema) -> list[str]:
        return sorted(
            name
            for name, scalar in schema.scalars.items()
            if scalar.has_custom_parse and not self.has(name)
        )


class ScalarNormalizeRegistry:
    """Registry of scalar normalize functions keyed by canonical name."""

    def __init__(self) -> None:
        self._fns: dict[str, ScalarNormalizeFunc] = {}

    def register(self, name: str, fn: ScalarNormalizeFunc) -> None:
        self._fns[name] = fn

    def unregister(self, name: str) -> bool:
        return self._fns.pop(name, None) is not None

    def get(self, name: str) -> ScalarNormalizeFunc | None:
        return self._fns.get(name)

    def has(self, name: str) -> bool:
        return name in self._fns

    def names(self) -> list[str]:
        return sorted(self._fns)

    def missing_normalizers(self, schema: Schema) -> list[str]:
        return sorted(
            name
            for name, scalar in schema.scalars.items()
            if scalar.has_custom_normalize and not self.has(name)
        )


def create_default_scalar_parse_registry() -> ScalarParseRegistry:
    """Build a parse registry pre-populated with every scalar that ships a
    custom ``parse_<name>``."""
    registry = ScalarParseRegistry()
    for canonical_name, fn in DEFAULT_PARSE_FUNCTIONS.items():
        # Human label for the error message: the segment after the dot.
        label = canonical_name.split(".")[-1] if "." in canonical_name else canonical_name
        registry.register(canonical_name, _wrap_parse(fn, label))
    return registry


def create_default_scalar_normalize_registry() -> ScalarNormalizeRegistry:
    """Build a normalize registry pre-populated with every scalar that ships a
    custom ``normalize_<name>``."""
    registry = ScalarNormalizeRegistry()
    for canonical_name, fn in DEFAULT_NORMALIZE_FUNCTIONS.items():
        registry.register(canonical_name, fn)
    return registry


__all__ = [
    "ScalarNormalizeFunc",
    "ScalarNormalizeRegistry",
    "ScalarParseFunc",
    "ScalarParseRegistry",
    "create_default_scalar_normalize_registry",
    "create_default_scalar_parse_registry",
]

File: runtime/schema/typescript/src/runtime/parse/registry.ts
/**
 * Scalar parse and normalize registries: map canonical scalar names (e.g.
 * "Contact.Email") to generated superscalar implementations. Validation has its
 * own registry in runtime/validation/registry.ts.
 */

import { backend } from 'superscalar/backend';
import * as generatedScalars from 'superscalar/scalars';
import { SCALAR_METADATA, scalarIdByCanonical } from 'superscalar/scalars';
import type { ValidationError } from 'superscalar/validation';
import { BUILTIN_SCALARS } from '../builtin-scalars.generated';
import type { Schema } from '../validation/types';

export type ScalarParseFunc = (input: string) => [string, ValidationError[]];
export type ScalarNormalizeFunc = (input: string) => string;

function normalizeScalarRegistryKey(name: string): string {
  const trimmed = name.trim();
  if (!trimmed || trimmed.includes('.')) {
    return trimmed;
  }
  return trimmed.includes('_') ? trimmed.split('_').join('.') : trimmed;
}

export class ScalarParseRegistry {
  private readonly fns = new Map<string, ScalarParseFunc>();

  constructor(initial?: Record<string, ScalarParseFunc>) {
    if (!initial) return;
    for (const [name, fn] of Object.entries(initial)) {
      this.fns.set(normalizeScalarRegistryKey(name), fn);
    }
  }

  register(name: string, fn: ScalarParseFunc): void {
    this.fns.set(normalizeScalarRegistryKey(name), fn);
  }

  unregister(name: string): boolean {
    return this.fns.delete(normalizeScalarRegistryKey(name));
  }

  get(name: string): ScalarParseFunc | undefined {
    return this.fns.get(normalizeScalarRegistryKey(name));
  }

  has(name: string): boolean {
    return this.fns.has(normalizeScalarRegistryKey(name));
  }

  names(): string[] {
    return [...this.fns.keys()].sort((a, b) => a.localeCompare(b));
  }

  missingParsers(schema: Schema): string[] {
    const missing: string[] = [];
    for (const scalar of Object.values(schema.scalars || {})) {
      if (scalar.hasCustomParse && !this.has(scalar.name)) {
        missing.push(scalar.name);
      }
    }
    return missing.sort((a, b) => a.localeCompare(b));
  }
}

export class ScalarNormalizeRegistry {
  private readonly fns = new Map<string, ScalarNormalizeFunc>();

  constructor(initial?: Record<string, ScalarNormalizeFunc>) {
    if (!initial) return;
    for (const [name, fn] of Object.entries(initial)) {
      this.fns.set(normalizeScalarRegistryKey(name), fn);
    }
  }

  register(name: string, fn: ScalarNormalizeFunc): void {
    this.fns.set(normalizeScalarRegistryKey(name), fn);
  }

  unregister(name: string): boolean {
    return this.fns.delete(normalizeScalarRegistryKey(name));
  }

  get(name: string): ScalarNormalizeFunc | undefined {
    return this.fns.get(normalizeScalarRegistryKey(name));
  }

  has(name: string): boolean {
    return this.fns.has(normalizeScalarRegistryKey(name));
  }

  names(): string[] {
    return [...this.fns.keys()].sort((a, b) => a.localeCompare(b));
  }

  missingNormalizers(schema: Schema): string[] {
    const missing: string[] = [];
    for (const scalar of Object.values(schema.scalars || {})) {
      if (scalar.hasCustomNormalize && !this.has(scalar.name)) {
        missing.push(scalar.name);
      }
    }
    return missing.sort((a, b) => a.localeCompare(b));
  }
}

function wrapParse(
  fn: (input: unknown) => unknown | null,
  label: string,
): ScalarParseFunc {
  return (input: string) => {
    const parsed = fn(input);
    if (parsed === null) {
      return [input, [{ validator: 'parse', message: `invalid ${label}` }]];
    }
    return [String(parsed), []];
  };
}

function wrapNormalize(
  fn: (input: unknown) => unknown | null,
): ScalarNormalizeFunc {
  return (input: string) => {
    const normalized = fn(input);
    if (normalized === null) {
      return input;
    }
    return String(normalized);
  };
}

// Route through the core and keep its canonical text. Temporal.DateTime gets
// the exact RFC3339Nano form (sub-seconds preserved/trimmed, numeric offset
// kept), not JS `toISOString()`, which forces `.000` padding and UTC. A
// JSON-shaped scalar (json_schema type `object`, such as Generic.StringMap)
// gets canonical JSON text: its generated parse function returns the decoded
// map, which wrapParse would stringify as `[object Object]`.
function coreTextAdapter(canonicalName: string, label: string): ScalarParseFunc {
  const id = scalarIdByCanonical[canonicalName];
  return (input: string) => {
    try {
      return [backend.parse(id, input), []];
    } catch {
      return [input, [{ validator: 'parse', message: `invalid ${label}` }]];
    }
  };
}

type GeneratedScalarFn = (input: unknown) => unknown | null;

/**
 * Looks up the generated superscalar `parse<Symbol>` / `normalize<Symbol>`
 * function for a canonical name, using the symbol the metadata table carries.
 */
function generatedScalarFn(prefix: 'parse' | 'normalize', symbol: string): GeneratedScalarFn | undefined {
  const candidate = (generatedScalars as Record<string, unknown>)[prefix + symbol];
  if (typeof candidate !== 'function') {
    return undefined;
  }
  return candidate as GeneratedScalarFn;
}

/**
 * Default parse registry: every scalar whose generated catalog entry
 * (builtin-scalars.generated.ts, written by internal/tools/scalarcatalog) marks a
 * custom parse step, adapted over the generated superscalar parse function. The
 * walker only consults the registry for hasCustomParse scalars, so those are
 * the only names registered; the set follows the catalog, not a hand list.
 * Temporal.DateTime and JSON-shaped scalars route through the core directly
 * and return its canonical text (coreTextAdapter).
 */
export function createDefaultScalarParseRegistry(): ScalarParseRegistry {
  const r = new ScalarParseRegistry();
  for (const meta of SCALAR_METADATA) {
    const def = BUILTIN_SCALARS[meta.canonicalName.replace(/\./g, '_')];
    if (!def || !def.hasCustomParse) continue;
    const label = meta.canonicalName.slice(meta.canonicalName.indexOf('.') + 1);
    if (meta.canonicalName === 'Temporal.DateTime' || def.typeMappings.json_schema === 'object') {
      r.register(meta.canonicalName, coreTextAdapter(meta.canonicalName, label));
      continue;
    }
    const fn = generatedScalarFn('parse', meta.symbol);
    if (!fn) continue;
    r.register(meta.canonicalName, wrapParse(fn, label));
  }
  return r;
}

/**
 * Default normalize registry: every scalar whose generated catalog entry marks
 * a custom normalize step, adapted over the generated superscalar normalize
 * function.
 */
export function createDefaultScalarNormalizeRegistry(): ScalarNormalizeRegistry {
  const r = new ScalarNormalizeRegistry();
  for (const meta of SCALAR_METADATA) {
    const def = BUILTIN_SCALARS[meta.canonicalName.replace(/\./g, '_')];
    if (!def || !def.hasCustomNormalize) continue;
    const fn = generatedScalarFn('normalize', meta.symbol);
    if (!fn) continue;
    r.register(meta.canonicalName, wrapNormalize(fn));
  }
  return r;
}
